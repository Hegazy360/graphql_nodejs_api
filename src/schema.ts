import { join } from 'path';
import "graphql-import-node";
import resolvers from "./resolvers";
import { loadSchemaSync } from '@graphql-tools/load';
import { addResolversToSchema } from '@graphql-tools/schema';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';

const schema = loadSchemaSync(join(__dirname, './schema/schema.graphql'), {
  loaders: [
    new GraphQLFileLoader(),
  ]
});

const schemaWithResolvers = addResolversToSchema({
  schema,
  resolvers,
});

export default schemaWithResolvers;
