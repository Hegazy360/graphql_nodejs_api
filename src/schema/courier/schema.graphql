# import Courier from "./courier.graphql"

type Query {
  getAllCouriers: [Courier!],
  lookUpCouriersByMaxCapacity(capacity_required: Int!): [Courier!] 
}

type Mutation {
  updateCourierCapacity(id: ID!, max_capacity: Int!): Courier,
  removeCourierCapacity(id: ID!): Courier,
}
