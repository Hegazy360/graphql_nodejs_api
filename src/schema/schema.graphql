# import * from "./courier/schema.graphql"
