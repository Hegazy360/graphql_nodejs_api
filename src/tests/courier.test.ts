import { createTestClient } from 'apollo-server-testing'
import { ApolloServer, gql } from "apollo-server-express";
// import CourierMocks from '../mocks/courier';
import schema from '../schema'

const server = new ApolloServer({
  schema,
  // Leaving this here as a demonstration of how we would add custom mocks
  // when testing and the architecture I'd follow in such case.
  // mocks: {
  //   ...CourierMocks
  // },
});

const GET_ALL_COURIERS = gql`
  query getAllCouriers {
    getAllCouriers {
      id
    }
  }
`;

const LOOKUP_COURIER_BY_CAPACITY = gql`
  query lookUpCouriersByMaxCapacity($capacity_required: Int!) {
    lookUpCouriersByMaxCapacity(capacity_required: $capacity_required) {
      id,
      max_capacity
    }
  }
`;

const UPDATE_COURIER_CAPACITY = gql`
  mutation updateCourierCapacity($id: ID!, $max_capacity: Int!) {
    updateCourierCapacity(id: $id, max_capacity: $max_capacity) {
      id,
      max_capacity
    }
  }
`;

const REMOVE_COURIER_CAPACITY = gql`
  mutation updateCourierCapacity($id: ID!) {
    removeCourierCapacity(id: $id) {
      id,
      max_capacity
    }
  }
`;

const { query, mutate } = createTestClient(server);

test("I can fetch all couriers", async () => {
  const res = await query({ query: GET_ALL_COURIERS });

  expect(res).toMatchSnapshot();
});

test("I can lookup couriers by capacity", async () => {
  const res = await query({
    query: LOOKUP_COURIER_BY_CAPACITY,
    variables: { capacity_required: 12 }
  });

  expect(res).toMatchSnapshot();
});

test("I can update a courier's max capacity", async () => {
  const res = await mutate({
    mutation: UPDATE_COURIER_CAPACITY,
    variables: { id: "1", max_capacity: 12 },
  });

  expect(res).toMatchSnapshot();
});

test("I can't update a courier's max capacity without passing an id", async () => {
  const res = await mutate({
    mutation: UPDATE_COURIER_CAPACITY,
    variables: { max_capacity: 12 },
  });

  expect(res).toMatchSnapshot();
});

test("I can't update a courier's max capacity without passing a valid id", async () => {
  const res = await mutate({
    mutation: UPDATE_COURIER_CAPACITY,
    variables: {
      id: "INVALID_ID",
      max_capacity: 12
    },
  });

  const { errors } = await res;

  expect((errors[0].extensions).code).toEqual("COURIER_NOT_FOUND")
  expect(res).toMatchSnapshot();
});

test("I can't update a courier's max capacity without passing a new maximum capacity", async () => {
  const res = await mutate({
    mutation: UPDATE_COURIER_CAPACITY,
    variables: { id: "1" },
  });

  expect(res).toMatchSnapshot();
});

test("I can remove a courier's max capacity", async () => {
  const res = await mutate({
    mutation: REMOVE_COURIER_CAPACITY,
    variables: { id: "1" },
  });

  expect(res).toMatchSnapshot();
});


