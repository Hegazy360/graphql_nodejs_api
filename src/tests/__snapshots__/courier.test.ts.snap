// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I can fetch all couriers 1`] = `
Object {
  "data": Object {
    "getAllCouriers": Array [
      Object {
        "id": "1",
      },
      Object {
        "id": "2",
      },
      Object {
        "id": "3",
      },
      Object {
        "id": "4",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`I can lookup couriers by capacity 1`] = `
Object {
  "data": Object {
    "lookUpCouriersByMaxCapacity": Array [
      Object {
        "id": "2",
        "max_capacity": 20,
      },
      Object {
        "id": "3",
        "max_capacity": 30,
      },
      Object {
        "id": "4",
        "max_capacity": 45,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`I can remove a courier's max capacity 1`] = `
Object {
  "data": Object {
    "removeCourierCapacity": Object {
      "id": "1",
      "max_capacity": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`I can update a courier's max capacity 1`] = `
Object {
  "data": Object {
    "updateCourierCapacity": Object {
      "id": "1",
      "max_capacity": 12,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`I can't update a courier's max capacity without passing a new maximum capacity 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [GraphQLError: Variable "$max_capacity" of required type "Int!" was not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`I can't update a courier's max capacity without passing a valid id 1`] = `
Object {
  "data": Object {
    "updateCourierCapacity": null,
  },
  "errors": Array [
    [GraphQLError: Courier not found.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`I can't update a courier's max capacity without passing an id 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [GraphQLError: Variable "$id" of required type "ID!" was not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
